src/shared/driverUtils.js

```diff
@@ -0,0 +1,258 @@
+const axios = require("axios");
+const { exec, execSync } = require("child_process");
+
+fluid.registerNamespace("fluid.personalData");
+fluid.personalData.startServerCmd = "node startServer.js";
+
+const NUM_CHECK_REQUESTS = 10;
+const DELAY = 2000; // msec
+
+fluid.personalData.sleep = async function (delay) {
+    return new Promise((resolve) => setTimeout(resolve, delay));
+};
+
+/**
+ * Check if the personal data server is running using the given url.  If not
+ * running, start it. If it does not start after ten attempts, give up.
+ *
+ * @param {Number} serverPort - The port that the server is listening to.
+ * @return {Object} An object that has the status of the last request regardless
+ *                  of success, and the node ChildProcess started by `exec()`
+ *                  or `null` if it failed to start.
+ */
+fluid.personalData.startServer = async function (serverPort) {
+    const url = "http://localhost:" + serverPort;
+    let resp = {};
+
+    console.debug(`- Checking if server is running using ${url}`);
+    try {
+        resp = await axios(url);
+        return { status: resp.status, process: null, wasRunning: true };
+    }
+    catch (error) { ; }
+
+    // Server needs to be started
+    console.debug(`- Starting server at ${url}`);
+    const pdServerProcess = exec(fluid.personalData.startServerCmd + " " + serverPort);
+    for (let i = 0; i < NUM_CHECK_REQUESTS; i++) {
+        try {
+            console.debug(`... attempt ${i}`);
+            resp = await axios(url);
+        }
+        catch (error) {
+            console.debug(`... attempt ${i} error: ${error.message}`);
+            resp.status = 503;
+        }
+        if (resp.status === 200) {
+            break;
+        } else {
+            console.debug(`... attempt ${i} going to sleep`);
+            await fluid.personalData.sleep(DELAY);
+        }
+    }
+    console.debug(`- Attempt to start server, result:  ${resp.status}`);
+    return { status: resp.status, process: pdServerProcess, wasRunning: false };
+};
+
+/**
+ * Stop the personal data server, if it was started by this testing sequence.
+ * If it was started in order to run the tests, the `childProcess` argument will
+ * have a non-null ChildProcess instance and a `wasRunning` flag set to false.
+ * If so, check the ChildProcess's exit status and, if it has not exited, invoke
+ * its `kill()` method.
+ *
+ * @param {Object} childProcessInfo - Information about the server's process
+ * @param {Object} childProcessInfo.wasRunning - True if the server was already
+ *                                               running.
+ * @param {Object} childProcessInfo.process - ChildProcess instance `wasRunning`
+ *                                            created if the server was not
+ *                                            running.
+ * @param {String} url - Url to use to check if the ChildProcess has been killed.
+ */
+fluid.personalData.stopServer = async function (childProcessInfo, url) {
```

Comment:
There seems to be some race condition or mishandling of status codes associated with this function - on my Ubuntu image it often goes through the 10 retries and then it seems quietly bails out:
````
07:43:46.079:  jq: Test concluded - Module "Personal Data Server /health and /ready tests." Test name "Health and Ready end point tests": 14/14 passed - PASS
- Starting database container PostgresForTests
Checking that PostgresDB is ready...
... attempt 0
- Attempt to start database, result:  true
Creating (or checking existence of) personalDataTests ...
createdb: error: database creation failed: ERROR:  database "personalDataTests" already exists
- Checking if server is running using http://localhost:3001
- Emptying database...
- ... defining the tables
- ... adding initial table records
- Sending '/ready' request
- Sending '/sso/google'
- Calling googleSso.fetchAccessToken(/token)
Status: 200: access token for google
- Calling googleSso.fetchAccessToken(/token)
- Calling googleSso.fetchUserInfo(/userInfo)
Status: 200: user profile for google
- Calling googleSso.fetchUserInfo(/userInfo)
- Calling googleSso.storeUserAndAccessToken()
...adding access token for PatId to database
- Sending '/sso/google/login/callback' request
- Deleting user with id 'PatId'
- Stopping database container PostgresForTests
- Stopping server at http://localhost:3001
... attempt 0
... attempt 1
... attempt 2
... attempt 3
... attempt 4
... attempt 5
... attempt 6
... attempt 7
... attempt 8
... attempt 9
07:44:08.946:  jq: Test concluded - Module "Personal Data Server Google SSO Tests" Test name "Google SSO tests": 36/36 passed - PASS
''''